apps/email/js/cards/cmp/autocomplete.js

```diff
@@ -0,0 +1,565 @@
+'use strict';
+define(function(require) {
+
+var GestureDetector = require('shared/js/gesture_detector'),
+    itemTemplate = require('tmpl!./autocomplete_item.html'),
+    mozL10n = require('l10n!'),
+    regExpEscape = require('regExpEscape'),
+    transitionEnd = require('transition_end'),
+    validTypes = {
+      email: true,
+      text: true
+    };
+
+/**
+ * Converts a text string into a span and some text nodes that are placed
+ * as the children of node. The span contains the match for the queryRegExp.
+ * @param  {RegExp} queryRegExp The regexp whose match is highlighted via a
+ * span in the target text.
+ * @param  {String} text        The target text to find the match.
+ * @param  {Element} node       The element that will display the matched text.
+ */
+function formatMatchString(queryRegExp, text, node) {
+  // Just highlight the first match.
+  var match = queryRegExp.exec(text);
+
+  if (match) {
+    var index = match.index,
+        endIndex = index + queryRegExp.source.length;
+
+    var startString = text.substring(0, index);
+    if (startString) {
+      node.appendChild(document.createTextNode(startString));
+    }
+
+    var matchString = text.substring(index, endIndex);
+    if (matchString) {
+      var span = document.createElement('span');
+      span.classList.add('highlight');
+      span.textContent = matchString;
+      node.appendChild(span);
+    }
+
+    var endString = text.substring(endIndex);
+    if (endString) {
+      node.appendChild(document.createTextNode(endString));
+    }
+  } else {
+    node.textContent = text;
+  }
+}
+
+/**
+ * Given a node, if it or an ancestor has the CSS className, return the node
+ * with that className. Returns null if not a match.
+ * @param  {Node} node
+ * @return {Node}
+ */
+function findAncestorWithClass(node, className) {
+  if (node.classList && node.classList.contains(className)) {
```

Comment:
Good to know, will remember that if we see a lot of this sort of ancestor work going on. Since autocomplete is using a CSS class based API to find things, I like the specialized behavior here, although I have updated it locally to use a for loop. Thanks for pointing that out, good to remove the redundancy.